File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        len_s = len(s)
        dp = [[False] * len_s for _ in range(len_s)]
        result = []
        self.dfs(result, s, 0, [], dp)
        return result

    def dfs(
        self,
        result: List[List[str]],
        s: str,
        start: int,
        currentList: List[str],
        dp: List[List[bool]],
    ):
        if start >= len(s):
            result.append(list(currentList))
        for end in range(start, len(s)):
            if s[start] == s[end] and (
                end - start <= 2 or dp[start + 1][end - 1]
            ):
                dp[start][end] = True
                currentList.append(s[start : end + 1])
                self.dfs(result, s, end + 1, currentList, dp)
                currentList.pop()
#class Solution:
#    def partition(self, s: str) -> List[List[str]]:
#        ans = []
#        mid = len(s) // 2
#        for i in range(1, len(s)+1):
#            temp = []
#            for j in range(0, len(s)):
#                start = s[j:j+i]
#                flag = True
#                for k in range(len(start)//2+1):
#                    if start[k] != start[-k-1]:
#                        flag = False
#                        break
#                if flag == True:
#                    temp.append(start)
#            ans.append(temp)
#        return ans
#

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> currVec;
        dfs(ans, s, 0, currVec);
        return ans; 
    }

    void dfs(vector<vector<string>>& res, string& s, int start, vector<string>& currVec) {
        if (start >= s.length()) res.push_back(currVec);
        for (int end = start; end < s.length(); end++) {
            if (isPalindrome(s, start, end)) {
                currVec.push_back(s.substr(start, end-start+1));
                dfs(res, s, end+1, currVec);
                currVec.pop_back();
            }
        }
    }

    bool isPalindrome(string& s, int low, int high) {
        while (low < high) {
            if (s[low++] != s[high--]) return false;
        }
        return true;
    }
};